src/Jaeger.Core/Reporters/RemoteReporter.cs

```diff
@@ -1,114 +1,222 @@
             try
             {
-                // TODO: This Task should be queued and be processed in a separate thread
-                await _transport.AppendAsync(span, CancellationToken.None).ConfigureAwait(false);
+                // Note: Java creates a CloseCommand but we have CompleteAdding() in C# so we don't need the command.
+                _commandQueue.CompleteAdding();
+
+                // Give sender some time to process any queued spans.
+                _queueProcessor.Wait(10000);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Dispose interrupted");
             }
-            catch (Exception e)
+            finally
             {
-                _logger.LogError(e, "Unable to report span in RemoteReporter");
-                _metrics.ReporterDropped.Inc(1);
+                _flushTimer.Dispose();
+                try
+                {
+                    int n = _sender.Close();
+                    _metrics.ReporterSuccess.Inc(n);
+                }
+                catch (SenderException ex)
+                {
+                    _metrics.ReporterFailure.Inc(ex.DroppedSpanCount);
+                }
             }
         }
 
-        // TODO: Make async!
-        private async void Flush()
+        internal void Flush()
         {
+            // to reduce the number of updateGauge stats, we only emit queue length on flush
+            _metrics.ReporterQueueLength.Update(_commandQueue.Count);
+
             try
             {
-                // TODO: Not exposed, this should be the list of unprocessed Report calls
-                //_metrics.ReporterQueueLength.Update(_commandQueue.Count);
-                int n = await _transport.FlushAsync(CancellationToken.None).ConfigureAwait(false);
-                _metrics.ReporterSuccess.Inc(n);
+                // We can safely drop FlushCommand when the queue is full - sender should take care of flushing
+                // in such case
+                _commandQueue.TryAdd(new FlushCommand(this));
             }
-            catch (SenderException e)
+            catch (InvalidOperationException)
             {
-                _logger.LogError(e, "Unable to flush RemoteReporter");
-                _metrics.ReporterFailure.Inc(e.DroppedSpans);
+                // The queue has been marked as Completed -> no-op.
+            }
+            catch (Exception ex)
+            {
+                _logger.LogWarning(ex, $"{nameof(Flush)}() failed");
             }
         }
 
-        public class Builder
+        /*
+         * The code below implements the command pattern.  This pattern is useful for
+         * situations where multiple threads would need to synchronize on a resource,
+         * but are fine with executing sequentially.  The advantage is simplified code where
+         * tasks are put onto a blocking queue and processed sequentially by another thread.
+         */
+        public interface ICommand
```

Comment:
My personal preference is to have stuff more separated as well, but I did it in the same way as Java because this way it's easier to compare files. The (potentially?!) better readability with multiple files was not worth deviating IMO.

I wonder how we should decide how to proceed here? Again, my vote is for being as close to Java as possible but I want to ensure we discuss this properly if you/anyone disagrees.

I'm writing this comment just once but it also applies to all other cases as well (e.g. Builder types).